exchanges/bot.go

```diff
@@ -722,3 +799,113 @@ func (bot *ExchangeBot) Conversion(dcrVal float64) *Conversion {
 	}
 	return nil
 }
+
+// Fetch the pre-encoded JSON chart data from the cache, if it exists and is not
+// expired. Data is considered expired only when newer data has been received,
+// but not necessarily requested/encoded yet. Boolean `hit` indicates success.
+func (bot *ExchangeBot) fetchFromCache(chartID string) (data []byte, bestVersion int, hit bool) {
+	bot.mtx.RLock()
+	defer bot.mtx.RUnlock()
+	bestVersion = bot.cachedChartVersion(chartID)
+	cache, found := bot.versionedCharts[chartID]
+	if found {
+		if cache.dataID == bestVersion {
+			return cache.chart, bestVersion, true
+		}
+	}
+	return data, bestVersion, false
+}
+
+// QuickSticks returns the up-to-date candlestick data for the specified
+// exchange and bin width, pulling from the cache if appropriate.
+func (bot *ExchangeBot) QuickSticks(token string, rawBin string) ([]byte, error) {
+	chartID := genCacheID(token, rawBin)
+	bin := candlestickKey(rawBin)
+	data, bestVersion, isGood := bot.fetchFromCache(chartID)
+	if isGood {
+		return data, nil
+	}
+
+	// No hit on cache. Re-encode.
+
+	bot.mtx.Lock()
+	defer bot.mtx.Unlock()
+	state, found := bot.currentState.DcrBtc[token]
+	if !found {
+		return []byte{}, fmt.Errorf("Failed to find DCR exchange state for %s", token)
```

Comment:
I was long under the impression that doing things like calling `len` on a nil slice would cause a panic. I don't know where I got that, but I realized my folly only recently.